test/unit/test_pretty.py

```diff
@@ -0,0 +1,142 @@
+
+
+def test_small():
+    status_repr = repr(MINIMAL_STATUS)
+    assert (
+        status_repr
+        == "Status(model=ModelStatus(name='mdl', type='typ', controller='ctl', cloud='aws', version='3.0.0'), machines={}, apps={})"
+    )
+    assert str(MINIMAL_STATUS) == status_repr
+
+    # Ensure it's actually a Python repr: it can roundtrip via eval()
+    assert eval(status_repr, jubilant.statustypes.__dict__) == MINIMAL_STATUS
+
+
+def test_medium():
+    status = jubilant.Status._from_dict(json.loads(SNAPPASS_JSON))
+    status_repr = repr(status)
+    assert (
+        status_repr
+        == """\
+Status(
+  model=ModelStatus(
+    name='tt',
+    type='caas',
+    controller='microk8s-localhost',
+    cloud='microk8s',
+    version='3.6.1',
+    region='localhost',
+    model_status=StatusInfoContents(current='available', since='24 Feb 2025 12:02:57+13:00'),
+  ),
+  machines={},
+  apps={
+    'snappass-test': AppStatus(
+      charm='snappass-test',
+      charm_origin='charmhub',
+      charm_name='snappass-test',
+      charm_rev=9,
+      exposed=False,
+      base=FormattedBase(name='ubuntu', channel='20.04'),
+      charm_channel='latest/stable',
+      scale=1,
+      provider_id='276bec9f-6a0c-46ea-8094-aca6337d46e5',
+      address='10.152.183.248',
+      app_status=StatusInfoContents(current='active', message='snappass started', since='24 Feb 2025 12:03:17+13:00'),
+      units={
+        'snappass-test/0': UnitStatus(
+          workload_status=StatusInfoContents(current='active', message='snappass started', since='24 Feb 2025 12:03:17+13:00'),
+          juju_status=StatusInfoContents(current='idle', since='24 Feb 2025 12:03:18+13:00', version='3.6.1'),
+          leader=True,
+          address='10.1.164.138',
+          provider_id='snappass-test-0',
+        ),
+      },
+    ),
+  },
+  controller=ControllerStatus(timestamp='12:04:55+13:00'),
+)"""
+    )
+    assert str(status) == status_repr
+    assert eval(status_repr, jubilant.statustypes.__dict__) == status
+
+
+def test_large():
+    status = jubilant.Status._from_dict(json.loads(DATABASE_WEBAPP_JSON))
+    status_repr = repr(status)
+    assert (
+        status_repr
+        == """\
+Status(
+  model=ModelStatus(
+    name='tt',
+    type='caas',
+    controller='microk8s-localhost',
+    cloud='microk8s',
+    version='3.6.1',
+    region='localhost',
+    model_status=StatusInfoContents(current='available', since='24 Feb 2025 12:02:57+13:00'),
+  ),
+  machines={},
+  apps={
+    'database': AppStatus(
+      charm='local:database-0',
+      charm_origin='local',
+      charm_name='database',
+      charm_rev=0,
+      exposed=False,
+      base=FormattedBase(name='ubuntu', channel='22.04'),
+      scale=1,
+      provider_id='fa764a56-2b71-4f7e-a6eb-b265f13adc4c',
+      address='10.152.183.228',
+      app_status=StatusInfoContents(current='active', message='relation-created: added new secret', since='24 Feb 2025 16:59:43+13:00'),
```

Comment:
I suppose that not parsing Juju timestamp is a deliberate choice.

FWIW I prefer strings to datetimes.

Reply:
Yep, datetime objects are a mixed blessing, so I decided to KISS for now. I don't think these will be treated/compared as datetimes much, if at all, so the user can do it if they need to.